concurrent/src/test/java/io/airlift/concurrent/TestLocks.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "exception thrown under a lock")
+    public void testLockingWithRuntimeException()
+    {
+        locking(new ReentrantLock(), () -> {
+            throw new RuntimeException("exception thrown under a lock");
+        });
+    }
+
+    @Test(expectedExceptions = IOException.class, expectedExceptionsMessageRegExp = "exception thrown under a lock")
+    public void testLockingWithCheckedException()
+            throws IOException
+    {
+        locking(new ReentrantLock(), () -> {
+            throw new IOException("exception thrown under a lock");
+        });
+    }
+
+    @Test
+    public void testLockingLocks()
+            throws Exception
+    {
+        Lock lock = new ReentrantLock();
+        int threads = 10;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        AtomicInteger underLock = new AtomicInteger(0);
+        AtomicInteger maxUnderLock = new AtomicInteger(0);
+
+        List<Future<?>> futures = IntStream.range(0, threads)
+                .mapToObj(i -> executor.submit(() -> {
+                    barrier.await();
+                    locking(lock, () -> {
+                        maxUnderLock.accumulateAndGet(underLock.incrementAndGet(), Integer::max);
+                        Thread.sleep(30);
+                        underLock.decrementAndGet();
+                    });
+
+                    return null;
+                }))
+                .collect(toImmutableList());
+
+        futures.forEach(MoreFutures::getFutureValue);
+        assertEquals(maxUnderLock.get(), 1, "More than one thread acquired the lock at the same time");
+    }
+
+    @Test
+    public void testLockingInterruptiblyCanBeInterrupted()
+            throws Exception
+    {
+        Lock lock = new ReentrantLock();
+        Exchanger<Thread> exchanger = new Exchanger<>();
+
+        Future<Object> future = executor.submit(() -> {
+            exchanger.exchange(Thread.currentThread());
+            // main thread is under lock now
+            lockingInterruptibly(lock, () -> {
+                throw new AssertionError("test background thread got the lock");
+            });
+            throw new AssertionError("lockingInterruptibly returned, InterruptedException was expected");
+        });
+
+        locking(lock, () -> {
+            Thread testBackgroundThread = exchanger.exchange(null, 5, SECONDS);
+            Thread.sleep(40); // give the thread chance to lock
+            testBackgroundThread.interrupt();
+            // we can release lock now, as Lock.lockInterruptibly throws interrupted even if lock is available
+        });
+
+        try {
+            future.get(5, SECONDS);
+            fail("ExecutionException was expected");
+        }
+        catch (ExecutionException expected) {
+            if (!(expected.getCause() instanceof InterruptedException)) {
```

Comment:
Hm.. I'd like to have `expected.getCause()` or `expected` attached when check fails.